### DISCUSSION TITLE
Help: tmux binary with bin-gem-node

### DISCUSSION THREAD

**Discussion by albertfgu** [zdharma-continuum/zinit#391] [Q&A] [answered]
I recently updated all my plugins to use bin-gem-node for the convenient shims. However, I'm having a lot of different issues with tmux.

The original line was
```
zinit ice from"gh-r" as"program" mv"tmux* -> tmux"
zinit load tmux/tmux
```
This still works fine. However, it seems like it's recommended to move to an updated syntax that uses the `for` syntax and `sbin` to create shims. For example, most of my other plugins now use the syntax
```
zinit for \
    from'gh-r' \
    as'null' \
    sbin'**/nvim -> nvim' \
  neovim/neovim
```

With `tmux`, I've tried all sorts of combinations like this and they always fail. For example, I tried modifying the original command with all variations of `as` and `sbin`, e.g.
```
zinit ice from"gh-r" as"null" sbin"**/tmux* -> tmux"
zinit load tmux/tmux
```
and they all give some form of error.

In the [Recipes for popular programs](https://github.com/zdharma-continuum/zinit/wiki/Recipes-for-popular-programs#tmux) Wiki, the command they give is
```
zinit for \
    as'null' \
    atclone'./configure --disable-utf8proc --prefix=$PWD' \
    atpull'%atclone' \
    extract'!' \
    from'gh-r' \
    id-as'tmux' \
    make'PREFIX=$PWD install'\
    nocompile \
    sbin \
    ver'latest' \
  @tmux/tmux
```
However, this immediately runs into an error with not finding the right version. Deleting the `ver'latest'` line fixes this. However then it compiles tmux from scratch, and I don't see the point of this over just grabbing the binary.

**Question 1**: What's the point of the complicated tmux recipe? Is there any advantage of compiling from scratch?
**Question 2**: What's the simplest way to get the tmux binary from GH-release with the bgn shim?

**Comment by vladdoster** [comment]
Hi @albertfgu, 

Those are good questions and I can provide answers since I have added all the recipes.

>  What's the point of the complicated TMUX recipe? Is there any advantage of compiling from scratch?

The recipe compiles Tmux because the binary provided does not support macOS. Additionally, compiling Tmux takes a relatively short time, so I chose to add a recipe that would work for more users. Finally, I couldn't find a repository with pre-compiled macOS binaries.

> What's the simplest way to get the tmux binary from GH-release with the bgn shim?

Let me get back to you on this in a day or two.

**Comment by albertfgu** [comment]
I see, thanks! However, as mentioned in the original post I thought that the command
```
zinit ice from"gh-r" as"program" mv"tmux* -> tmux"
zinit load tmux/tmux
```
does actually still work and grab a correct binary on macOS. It just doesn't use the sbin shim, which I can't figure out how to do.

**Reply by vladdoster** [reply]
But the release asset does not contain a tmux binary?

**Comment by vladdoster** [comment, marked as answer]
@albertfgu, We recently added a `configure` ice, so this works:

```zsh
zinit for \
    as'null' \
    configure'--disable-utf8proc --prefix=$PWD --quiet' \
    make'PREFIX=$PWD --quiet install'\
    sbin \
  @tmux/tmux
```

<img width="1712" alt="Screenshot 2022-11-06 at 10 56 28" src="https://user-images.githubusercontent.com/10052309/200183977-14f703d1-5e60-402e-b447-9ae39635fe8a.png">

**Comment by vladdoster** [comment]
@albertfgu Pinging

**Comment by albertfgu** [comment]
Thanks for the response and sorry for the delay in following up! On one of my computers, the new command works beautifully. Thanks a lot!

On my other computer however (macOS Monterey 12.6.1, it's still being set up but I'm not sure which programs if any are missing...), I'm running into an error:
```
Downloading tmux/tmux…
Cloning into '/Users/albertgu/.local/share/zinit/plugins/tmux---tmux'...
⠙ ███████████ OBJ: 100, PACK: 46770/46770, COMPR: 100%, REC: 100%, RES: 100%
Running ./autogen.sh…
configure.ac:45: installing 'etc/compile'
configure.ac:10: installing 'etc/config.guess'
configure.ac:10: installing 'etc/config.sub'
configure.ac:8: installing 'etc/install-sh'
configure.ac:8: installing 'etc/missing'
Makefile.am: installing 'etc/depcomp'
configure.ac: installing 'etc/ylwrap'
configure.ac:47: warning: The macro `AC_PROG_CC_C99' is obsolete.
configure.ac:47: You should run autoupdate.
./lib/autoconf/c.m4:1659: AC_PROG_CC_C99 is expanded from...
configure.ac:47: the top level
Running ./configure --prefix=/Users/albertgu/.local/share/zinit/polaris…
./configure: line 4970: PKG_PROG_PKG_CONFIG: command not found
./configure: line 6079: syntax error near unexpected token `newline'
./configure: line 6079: `PKG_CHECK_MODULES('
Warning: zinit-configure-hook hook returned with 2
ERROR: No Makefile nor meson build dir found, make/meson isn't run!
```
I also noticed that on this computer, `which configure` points to `~/.local/share/zinit/polaris/bin/configure` whereas on the computer which works, `which configure` shows up empty. I'm not entirely sure what's going on. Thanks in advance!

**Comment by vladdoster** [comment]
@albertfgu I think youre missing `pkg-config`.

Try running:

```zsh
sudo xcode-select --install
```

It'll install or output the following:

```zsh
ᐳ sudo xcode-select --install  
Password:
xcode-select: error: command line tools are already installed, use "Software Update" to install updates
```

If you get the above, if you use Homebrew, you can run:

```zsh
brew install pkg-config
```

**Comment by albertfgu** [comment]
I followed that and also had to `brew install libevent` and everything works! Thanks so much for your help!!

**Comment by vladdoster** [comment]
Glad it works for you now. Thanks for using Zinit!

**Comment by albertfgu** [comment]
I recently tried installing on a new Linux machine and ran into the following output:
```
Downloading tmux/tmux…
Cloning into '/home/user/.local/share/zinit/plugins/tmux---tmux'...
⠼ ███████████ OBJ: 100, PACK: 47007/47007, COMPR: 100%, REC: 100%, RES: 100%
Running ./autogen.sh…
configure.ac:45: installing 'etc/compile'
configure.ac:10: installing 'etc/config.guess'
configure.ac:10: installing 'etc/config.sub'
configure.ac:8: installing 'etc/install-sh'
configure.ac:8: installing 'etc/missing'
Makefile.am: installing 'etc/depcomp'
configure.ac: installing 'etc/ylwrap'
Running ./configure --prefix=/home/user/.local/share/zinit/polaris…
Running `make PREFIX=$PWD --quiet install`…
format.c: In function ‘format_find’:
format.c:3542:5: warning: format not a string literal, format string not checked [-Wformat-nonliteral]
 3542 |     strftime(s, sizeof s, time_format, &tm);
      |     ^~~~~~~~
format.c: In function ‘format_expand1’:
format.c:4720:3: warning: format not a string literal, format string not checked [-Wformat-nonliteral]
 4720 |   if (strftime(expanded, sizeof expanded, fmt, &es->tm) == 0) {
      |   ^~
hyperlinks.c: In function ‘hyperlinks_put’:
hyperlinks.c:160:34: warning: format ‘%llX’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
  160 |  xasprintf(&external_id, "tmux%llX", hyperlinks_next_external_id++);
      |                               ~~~^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                  |                              |
      |                                  long long unsigned int         uint64_t {aka long unsigned int}
      |                               %lX
input.c: In function ‘input_reply_clipboard’:
input.c:2844:50: warning: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
 2844 |   if ((outlen = b64_ntop(buf, len, out, outlen)) == -1) {
      |                                                  ^~
notify.c: In function ‘notify_add’:
notify.c:196:36: warning: operand of ‘?:’ changes signedness from ‘int’ to ‘u_int’ {aka ‘unsigned int’} due to unsignedness of other operand [-Wsign-compare]
  196 |  ne->pane = (wp != NULL ? wp->id : -1);
      |                                    ^~
notify.c: In function ‘notify_hook’:
notify.c:243:51: warning: operand of ‘?:’ changes signedness from ‘int’ to ‘u_int’ {aka ‘unsigned int’} due to unsignedness of other operand [-Wsign-compare]
  243 |  ne.pane = (target->wp != NULL ? target->wp->id : -1);
      |                                                   ^~
tty-keys.c: In function ‘tty_keys_next’:
tty-keys.c:1211:6: warning: ‘terminator’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 1211 |  end -= terminator - 1;
      |  ~~~~^~~~~~~~~~~~~~~~~
tty-keys.c:1162:17: note: ‘terminator’ was declared here
 1162 |  size_t    end, terminator, needed;
      |                 ^~~~~~~~~~
tty-term.c: In function ‘tty_term_read_list’:
tty-term.c:731:3: warning: ‘s’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  731 |   xasprintf(&(*caps)[*ncaps], "%s=%s", ent->name, s);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 /usr/bin/mkdir -p 'WD/bin'
  /usr/bin/install -c tmux 'WD/bin'
```

All the output messages are warnings, so I would have expected it to still compile, but it seems to hang here. Any ideas?

**Comment by albertfgu** [comment]
Sorry for not clarifying, I used the exact same command that @vladdoster showed above:
```
zinit for \
    as'null' \
    configure'--disable-utf8proc --prefix=$PWD --quiet' \
    make'PREFIX=$PWD --quiet install'\
    sbin \
  @tmux/tmux
```

**Reply by vladdoster** [reply]
@albertfgu looking into this

**Comment by vladdoster** [comment]
@albertfgu,

Try the following:

```zsh
zinit for \
    as'null' \
    configure"--disable-utf8proc --prefix=$ZINIT[PLUGINS_DIR]/tmux --quiet" \
    id-as \
    make"PREFIX=$ZINIT[PLUGINS_DIR]/tmux --quiet install"\
    lbin'!' \
  @tmux/tmux
```

---

Context about the ices used:

1. `id-as` is used to make the location of the plugin deterministic. In this case, `tmux`.
2. Set prefix to the location of the plugin that you want to compile.
3. Profit?

If you are interested, @psprint and I don't quite agree on whether to use the plugin dir or $ZPFX as the prefix. See #458 for details.

IMO, it makes cleanup much more straightforward and ensures $ZPFX is not polluted with random compilation artifacts. Chime in if you have any opinions.

---

<img width="1160" alt="Screenshot 2023-04-28 at 19 41 13" src="https://user-images.githubusercontent.com/10052309/235273594-4a039f51-da26-4f20-a8e9-8b9ec83ab7ce.png">

**Reply by vladdoster** [reply]
Context about the ices used:

1. `id-as` is used to make the location of the plugin deterministic. In this case, `tmux`.
2. Set prefix to the location of the plugin that you want to compile.
3. Profit?

If you are interested, @psprint and I don't quite agree on whether to use the plugin dir or $ZPFX as the prefix. See #458 for details.

IMO, it makes cleanup much more straightforward and ensures $ZPFX is not polluted with random compilation artifacts. Chime in if you have any opinions.

**Comment by albertfgu** [comment]
I tried this command on a new machine and ran into the following message, which seems to have crashed my connection to the machine as well
```
Downloading lbin!… (at label: lbin!…)
Cloning into '/home/albertgu/.local/share/zinit/plugins/lbin!'...
remote: Not Found
fatal: repository 'https://github.com/lbin!/' not found

Clone failed (code: 128).

Downloading tmux/tmux…
Cloning into '/home/albertgu/.local/share/zinit/plugins/tmux---tmux'...
⠧ ███████████ OBJ: 100, PACK: 47093/47093, COMPR: 100%, REC: 100%, RES: 100%
==> Compiling autogen.sh [OK]
aclocal: error: 'configure.ac' is required
aclocal failed
Connection to <machine> closed.
```

**Comment by vladdoster** [comment]
@albertfgu,

As of v3.13.0, `tmux` can be installed by running:

```zsh
if [[ $OSTYPE =~ 'darwin*' ]]; then
    zinit configure'--disable-utf8proc' make for @tmux/tmux
else
    zinit build for @tmux/tmux
fi
```
